test/helper/common.bash

```diff
@@ -117,131 +232,207 @@ function repeat_in_container_until_success_or_timeout() {
-function private_config_path() {
-  echo "${PWD}/test/duplicate_configs/${1:-$(basename "${BATS_TEST_FILENAME}")}"
+  _exec_in_container /usr/bin/supervisorctl status "${SERVICE_NAME}" | grep 'RUNNING' &>/dev/null
 }
 
-function container_has_service_running() {
-  local CONTAINER_NAME="${1}"
-  local SERVICE_NAME="${2}"
-
-  docker exec "${CONTAINER_NAME}" /usr/bin/supervisorctl status "${SERVICE_NAME}" | grep RUNNING >/dev/null
+# Wait until a service is running.
+#
+# @param ${1} = container name
+# @param ${2} = name of the service to wait for
+function _wait_for_service() {
+  local CONTAINER_NAME=${1:?Container name must be provided}
+  local SERVICE_NAME="${2:?Service name must be provided}"
+
+  _repeat_until_success_or_timeout \
+    --fatal-test "_container_is_running ${CONTAINER_NAME}" \
+    "${TEST_TIMEOUT_IN_SECONDS}" \
+    _container_has_service_running "${CONTAINER_NAME}" "${SERVICE_NAME}"
 }
 
-function wait_for_service() {
-  local CONTAINER_NAME="${1}"
-  local SERVICE_NAME="${2}"
-
-  repeat_until_success_or_timeout --fatal-test "container_is_running ${CONTAINER_NAME}" "${TEST_TIMEOUT_IN_SECONDS}" \
-    container_has_service_running "${CONTAINER_NAME}" "${SERVICE_NAME}"
-}
 
+# TODO documentation @polarathene
+#
 # NOTE: Relies on ENV `LOG_LEVEL=debug` or higher
+#
+# @param ${1} = container name
+# @param ${2} = expected count [OPTIONAL]
 function _wait_until_expected_count_is_matched() {
   function __get_count() {
-    # NOTE: `|| true` required due to `set -e` usage:
-    # https://github.com/docker-mailserver/docker-mailserver/pull/2997#discussion_r1070583876
-    docker exec "${CONTAINER_NAME}" grep --count "${MATCH_CONTENT}" "${MATCH_IN_LOG}" || true
+    _exec_in_container grep --count "${MATCH_CONTENT}" "${MATCH_IN_LOG}"
   }
 
   # WARNING: Keep in mind it is a '>=' comparison.
   # If you provide an explict count to match, ensure it is not too low to cause a false-positive.
   function __has_expected_count() {
+    # shellcheck disable=2317
     [[ $(__get_count) -ge "${EXPECTED_COUNT}" ]]
   }
 
-  local CONTAINER_NAME=${1}
-  local EXPECTED_COUNT=${2}
-
-  # Ensure early failure if arg is missing:
-  assert_not_equal "${CONTAINER_NAME}" ''
+  local CONTAINER_NAME=${1:?Container name must be provided}
+  local EXPECTED_COUNT=${2:-}
 
   # Ensure the container is configured with the required `LOG_LEVEL` ENV:
   assert_regex \
-    $(docker exec "${CONTAINER_NAME}" env | grep '^LOG_LEVEL=') \
+    "$(_exec_in_container env | grep '^LOG_LEVEL=')" \
     '=(debug|trace)$'
 
   # Default behaviour is to wait until one new match is found (eg: incremented),
   # unless explicitly set (useful for waiting on a min count to be reached):
-  if [[ -z $EXPECTED_COUNT ]]
-  then
-    # +1 of starting count:
-    EXPECTED_COUNT=$(( $(__get_count) + 1 ))
-  fi
+  #
+  #                             +1 of starting count:
+  [[ -z ${EXPECTED_COUNT} ]] && EXPECTED_COUNT=$(bc <<< "$(__get_count) + 1")
```

Comment:
I don't think the comment indentation is necessary?

I see you have reverted the increment logic here to an earlier attempt I had (_or this was written by you previously and applied without updating_), is this intentional?

I'm not fussed with either approach personally, but I don't find the `|| true`  in `__get_count()` with the workaround comment to be any better (_if anything more awkward?_). @casperklein might have some input if there is a reason to prefer it though.

Reply:
For reference, explained here: https://github.com/docker-mailserver/docker-mailserver/pull/2997#discussion_r1070583876

From a logical point of view, `__get_count` shouldn't exit with code 1, if the function works as expected for our tests. With `bc`, this is not triggered, but the function still return with exit code 1. This is why I consider my solution as more accurate.

However, at the moment you two are primary working on/with the tests. That said, when you both are more comfortable with something else, I am fine with it 👍

Reply:
Done